# Reccomendation System for Restaurant Customer

- Author    : Muhammad Aditya Bayhaqie
- Practice  : Machine Learning Terapan (Dicoding)
- Dataset   : [Restaurant and Customer UCI](https://archive.ics.uci.edu/dataset/232/restaurant+consumer+data)

## Data Understanding

Let's get our Data and Library!!

In [16]:
import pandas as pd
import numpy as np

In [1]:
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/00232/RCdata.zip

--2025-04-03 08:52:04--  https://archive.ics.uci.edu/ml/machine-learning-databases/00232/RCdata.zip
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified
Saving to: ‘RCdata.zip’

RCdata.zip              [ <=>                ]  38.74K  --.-KB/s    in 0.009s  

2025-04-03 08:52:04 (4.09 MB/s) - ‘RCdata.zip’ saved [39674]



In [2]:
!unzip /content/RCdata.zip

Archive:  /content/RCdata.zip
  inflating: chefmozaccepts.csv      
  inflating: chefmozcuisine.csv      
  inflating: chefmozhours4.csv       
  inflating: chefmozparking.csv      
  inflating: geoplaces2.csv          
  inflating: rating_final.csv        
  inflating: README                  
  inflating: usercuisine.csv         
  inflating: userpayment.csv         
  inflating: userprofile.csv         


Kesembilan file (Mengecualikan README) tersebut dapat kita kategorikan menjadi 3, yaitu restaurant, consumers, dan user-item-rating.

### Data Assessment

In [6]:
accepts = pd.read_csv('/content/chefmozaccepts.csv')
cuisine = pd.read_csv('/content/chefmozcuisine.csv')
hours = pd.read_csv('/content/chefmozhours4.csv')
parking = pd.read_csv('/content/chefmozparking.csv')
geo = pd.read_csv('/content/geoplaces2.csv', encoding = "ISO-8859-1")
usercuisine = pd.read_csv('/content/usercuisine.csv')
payment = pd.read_csv('/content/userpayment.csv')
profile = pd.read_csv('/content/userprofile.csv')
rating = pd.read_csv('/content/rating_final.csv')

print('Jumlah data pembayaran yang diterima restoran: ', len(accepts.placeID.unique()))
print('Jumlah data masakan pada restoran: ', len(cuisine.placeID.unique()))
print('Jumlah data waktu buka restoran: ', len(hours.placeID.unique()))
print('Jumlah data lokasi restoran: ', len(geo.placeID.unique()))
print('Jumlah data masakan pengguna: ', len(usercuisine.userID.unique()))
print('Jumlah data profil pengguna: ', len(profile.userID.unique()))
print('Jumlah data penilaian yang diberikan pengguna: ', len(rating.userID.unique()))
print('Jumlah data penilaian restoran: ', len(rating.placeID.unique()))

Jumlah data pembayaran yang diterima restoran:  615
Jumlah data masakan pada restoran:  769
Jumlah data waktu buka restoran:  694
Jumlah data lokasi restoran:  130
Jumlah data masakan pengguna:  138
Jumlah data profil pengguna:  138
Jumlah data penilaian yang diberikan pengguna:  138
Jumlah data penilaian restoran:  130


## Exploratory Data Analysis

### Univariate Exploratory Data Analysis

`Accept` Variable

In [7]:
accepts.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1314 entries, 0 to 1313
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   placeID   1314 non-null   int64 
 1   Rpayment  1314 non-null   object
dtypes: int64(1), object(1)
memory usage: 20.7+ KB


In [8]:
print('Banyak data: ', len(accepts.placeID.unique()))
print('Jenis Pembayaran yang diterima: ', accepts.Rpayment.unique())

Banyak data:  615
Jenis Pembayaran yang diterima:  ['cash' 'VISA' 'MasterCard-Eurocard' 'American_Express' 'bank_debit_cards'
 'checks' 'Discover' 'Carte_Blanche' 'Diners_Club' 'Visa'
 'Japan_Credit_Bureau' 'gift_certificates']


Terdapat 615 data restoran yang unik dengan 12 jenis pembayaran yang diterima

`Cuisine` Variable

In [9]:
cuisine.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 916 entries, 0 to 915
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   placeID   916 non-null    int64 
 1   Rcuisine  916 non-null    object
dtypes: int64(1), object(1)
memory usage: 14.4+ KB


In [10]:
print('Banyak tipe masakan: ', len(cuisine.Rcuisine.unique()))
print('Tipe masakan: ', cuisine.Rcuisine.unique())

Banyak tipe masakan:  59
Tipe masakan:  ['Spanish' 'Italian' 'Latin_American' 'Mexican' 'Fast_Food' 'Burgers'
 'Dessert-Ice_Cream' 'Hot_Dogs' 'Steaks' 'Asian' 'International'
 'Mongolian' 'Vegetarian' 'Brazilian' 'Cafe-Coffee_Shop' 'Cafeteria'
 'Contemporary' 'Deli-Sandwiches' 'Diner' 'Japanese' 'Sushi' 'Seafood'
 'Chinese' 'Bar' 'Bar_Pub_Brewery' 'Pizzeria' 'Mediterranean' 'American'
 'Family' 'Caribbean' 'African' 'Breakfast-Brunch' 'Regional' 'Afghan'
 'Bakery' 'Game' 'Armenian' 'Vietnamese' 'Korean' 'Thai' 'Barbecue'
 'Polish' 'Dutch-Belgian' 'French' 'German' 'Southwestern' 'Persian'
 'Ethiopian' 'Juice' 'Soup' 'Continental-European' 'Greek' 'Southern'
 'Eastern_European' 'California' 'Bagels' 'Turkish' 'Organic-Healthy'
 'Fine_Dining']


Terdapat 59 tipe masakan yang berbeda dengan nama masakan seperti terlihat pada output kode.

`Profile` Variables

In [11]:
print(profile.shape)

(138, 19)


In [12]:
profile.head()

userID   latitude   longitude smoker     drink_level dress_preference  \
0  U1001  22.139997 -100.978803  false      abstemious         informal   
1  U1002  22.150087 -100.983325  false      abstemious         informal   
2  U1003  22.119847 -100.946527  false  social drinker           formal   
3  U1004  18.867000  -99.183000  false      abstemious         informal   
4  U1005  22.183477 -100.959891  false      abstemious    no preference   

  ambience transport marital_status        hijos  birth_year    interest  \
0   family   on foot         single  independent        1989     variety   
1   family    public         single  independent        1990  technology   
2   family    public         single  independent        1989        none   
3   family    public         single  independent        1940     variety   
4   family    public         single  independent        1992        none   

           personality  religion      activity  color  weight  budget  height  
0    thrifty-protector      none       student  black      69  medium    1.77  
1  hunter-ostentatious  Catholic       student    red      40     low    1.87  
2          hard-worker  Catholic       student   blue      60     low    1.69  
3          hard-worker      none  professional  green      44  medium    1.53  
4    thrifty-protector  Catholic       student  black      65  medium    1.69

Kita mendapatkan berbagai data mulai dari marital_status, birth_year, smoker, religion, budget, dll. Jika kita ingin membangun sistem rekomendasi yang lebih kompleks, mungkin kita perlu menyertakan fitur-fitur di atas pada model. Data profile ini berguna jika kita ingin membuat sistem rekomendasi berdasarkan demografi atau kondisi sosial pengguna.

`Rating` Variables

In [13]:
rating.head()

userID  placeID  rating  food_rating  service_rating
0  U1077   135085       2            2               2
1  U1077   135038       2            2               1
2  U1077   132825       2            2               2
3  U1077   135060       1            2               2
4  U1068   135104       1            1               2

In [14]:
rating.describe()

placeID       rating  food_rating  service_rating
count    1161.000000  1161.000000  1161.000000     1161.000000
mean   134192.041344     1.199828     1.215332        1.090439
std      1100.916275     0.773282     0.792294        0.790844
min    132560.000000     0.000000     0.000000        0.000000
25%    132856.000000     1.000000     1.000000        0.000000
50%    135030.000000     1.000000     1.000000        1.000000
75%    135059.000000     2.000000     2.000000        2.000000
max    135109.000000     2.000000     2.000000        2.000000

Dari output di atas, diketahui bahwa nilai maksimum rating adalah 2 dan nilai minimumnya adalah 0. Artinya, skala rating berkisar antara 0 hingga 2.

In [15]:
print('Jumlah userID: ', len(rating.userID.unique()))
print('Jumlah placeID: ', len(rating.placeID.unique()))
print('Jumlah data rating: ', len(rating))

Jumlah userID:  138
Jumlah placeID:  130
Jumlah data rating:  1161


## Data Preprocessing

`Restaurant` Merging

Untuk mempermudah model dalam mengolah data, kita akan menggabungkan data untuk mempermudah pemrosesannya nantinya

In [17]:
# Menggabungkan seluruh placeID pada kategori Restaurant
resto_all = np.concatenate((
    accepts.placeID.unique(),
    cuisine.placeID.unique(),
    hours.placeID.unique(),
    parking.placeID.unique(),
    geo.placeID.unique()
))

# Mengurutkan data dan menghapus data yang sama
resto_all = np.sort(np.unique(resto_all))

print('Jumlah seluruh data restoran berdasarkan placeID: ', len(resto_all))

Jumlah seluruh data restoran berdasarkan placeID:  938


Setelah file pada kategori Restaurant digabung, kita memiliki 938 restoran yang unik.

`User` Merging

Untuk mempermudah model dalam mengolah data, kita akan menggabungkan data untuk mempermudah pemrosesannya nantinya

In [18]:
# Menggabungkan seluruh userID
user_all = np.concatenate((
    usercuisine.userID.unique(),
    payment.userID.unique(),
    profile.userID.unique()
))

# Menghapus data yang sama kemudian mengurutkannya
user_all = np.sort(np.unique(user_all))

print('Jumlah seluruh user: ', len(user_all))

Jumlah seluruh user:  138


Kita hanya memiliki 138 data pengguna dari 938 restoran yang memiliki rating. Jauh lebih sedikit dibandingkan Restoran.

`Rating` Merging

Untuk mempermudah model dalam mengolah data, kita akan menggabungkan data untuk mempermudah pemrosesannya nantinya

In [19]:
# Menggabungkan file accepts, geo, parking, hours ke dalam dataframe resto_info
resto_info = pd.concat([accepts, geo, parking, hours])

# Menggabungkan dataframe rating dengan resto_info berdasarkan nilai placeID
resto = pd.merge(rating, resto_info , on='placeID', how='left')
resto

userID  placeID  rating  food_rating  service_rating Rpayment   latitude  \
0     U1077   135085       2            2               2     cash        NaN   
1     U1077   135085       2            2               2      NaN  22.150802   
2     U1077   135085       2            2               2      NaN        NaN   
3     U1077   135085       2            2               2      NaN        NaN   
4     U1077   135085       2            2               2      NaN        NaN   
...     ...      ...     ...          ...             ...      ...        ...   
8073  U1068   132660       0            0               0      NaN  23.752943   
8074  U1068   132660       0            0               0      NaN        NaN   
8075  U1068   132660       0            0               0      NaN        NaN   
8076  U1068   132660       0            0               0      NaN        NaN   
8077  U1068   132660       0            0               0      NaN        NaN   

       longitude                                     the_geom_meter  \
0            NaN                                                NaN   
1    -100.982680  0101000020957F00009F823DA6094858C18A2D4D37F9A4...   
2            NaN                                                NaN   
3            NaN                                                NaN   
4            NaN                                                NaN   
...          ...                                                ...   
8073  -99.164679  0101000020957F00003D7905C9DC8157C13FCD1AB7334E...   
8074         NaN                                                NaN   
8075         NaN                                                NaN   
8076         NaN                                                NaN   
8077         NaN                                                NaN   

                                       name  ...     accessibility   price  \
0                                       NaN  ...               NaN     NaN   
1                    Tortas Locas Hipocampo  ...  no_accessibility  medium   
2                                       NaN  ...               NaN     NaN   
3                                       NaN  ...               NaN     NaN   
4                                       NaN  ...               NaN     NaN   
...                                     ...  ...               ...     ...   
8073  carnitas mata calle Emilio Portes Gil  ...        completely     low   
8074                                    NaN  ...               NaN     NaN   
8075                                    NaN  ...               NaN     NaN   
8076                                    NaN  ...               NaN     NaN   
8077                                    NaN  ...               NaN     NaN   

      url Rambience franchise    area other_services parking_lot  \
0     NaN       NaN       NaN     NaN            NaN         NaN   
1       ?  familiar         f  closed           none         NaN   
2     NaN       NaN       NaN     NaN            NaN      public   
3     NaN       NaN       NaN     NaN            NaN         NaN   
4     NaN       NaN       NaN     NaN            NaN         NaN   
...   ...       ...       ...     ...            ...         ...   
8073    ?  familiar         f  closed           none         NaN   
8074  NaN       NaN       NaN     NaN            NaN        none   
8075  NaN       NaN       NaN     NaN            NaN         NaN   
8076  NaN       NaN       NaN     NaN            NaN         NaN   
8077  NaN       NaN       NaN     NaN            NaN         NaN   

             hours                  days  
0              NaN                   NaN  
1              NaN                   NaN  
2              NaN                   NaN  
3     00:00-00:00;  Mon;Tue;Wed;Thu;Fri;  
4     00:00-00:00;                  Sat;  
...            ...                   ...  
8073           NaN                   NaN  
8074           NaN                   NaN  
8075  00:00-23:30;  Mon;Tue;Wed;Thu;Fr

Kita memiliki banyak sekali missing value di sini. Mari kita cek lebih lanjut dengan kode berikut.

In [20]:
# Cek missing value dengan fungsi isnull()
resto.isnull().sum()

userID               0
placeID              0
rating               0
food_rating          0
service_rating       0
Rpayment          5781
latitude          6917
longitude         6917
the_geom_meter    6917
name              6917
address           6917
city              6917
state             6917
country           6917
fax               6917
zip               6917
alcohol           6917
smoking_area      6917
dress_code        6917
accessibility     6917
price             6917
url               6917
Rambience         6917
franchise         6917
area              6917
other_services    6917
parking_lot       6917
hours             4619
days              4619
dtype: int64

Hanya fitur userID, placeID, rating, food_rating, dan service_rating saja yang memiliki 0 missing value. Selanjutnya, mari kita hitung jumlah rating, food_rating, dan service rating berdasarkan place.ID

In [21]:
# Menghitung jumlah rating, food_rating, dan service kemudian menggabungkannya berdasarkan placeID
resto.groupby('placeID').sum()

userID  rating  \
placeID                                                              
132560   U1067U1067U1067U1067U1067U1067U1082U1082U1082U...      12   
132561   U1026U1026U1026U1026U1026U1129U1129U1129U1129U...      15   
132564   U1060U1060U1060U1060U1060U1080U1080U1080U1080U...      25   
132572   U1108U1108U1108U1108U1108U1108U1055U1055U1055U...      90   
132583   U1044U1044U1044U1044U1044U1044U1118U1118U1118U...      24   
...                                                    ...     ...   
135088   U1044U1044U1044U1044U1044U1044U1030U1030U1030U...      36   
135104   U1068U1068U1068U1068U1068U1068U1068U1068U1067U...      48   
135106   U1055U1055U1055U1055U1055U1055U1055U1055U1126U...      96   
135108   U1088U1088U1088U1088U1088U1126U1126U1126U1126U...      65   
135109   U1030U1030U1030U1030U1030U1030U1020U1020U1020U...      24   

         food_rating  service_rating  \
placeID                                
132560            24               6   
132561            20              20   
132564            25              30   
132572            90              84   
132583            24              30   
...              ...             ...   
135088            42              36   
135104            80              48   
135106            96              96   
135108            65              55   
135109            30              18   

                                                  Rpayment    latitude  \
placeID                                                                  
132560                                    cashcashcashcash   95.009216   
132561                                                   0   94.907276   
132564                                                   0   94.923698   
132572   cashcashcashcashcashcashcashcashcashcashcashca...  332.124707   
132583   cashVISAMasterCard-Eurocardbank_debit_cardscas...   75.689162   
...                                                    ...         ...   
135088                            cashcashcashcashcashcash  113.256068   
135104   cashVISAMasterCard-EurocardcashVISAMasterCard-...  166.270875   
135106   cashVISAMasterCard-EurocardcashVISAMasterCard-...  221.497088   
135108                                                   0  243.498787   
135109                                    cashcashcashcash   75.687139   

           longitude                                     the_geom_meter  \
placeID                                                                   
132560   -396.667653  0101000020957F0000FC60BDA8E88157C1B2C357D6DA4E...   
132561   -396.506024  0101000020957F000004457BB7AA8657C15F10835CD944...   
132564   -396.580739  0101000020957F0000EA4F00C5A08557C140085474D949...   
132572  -1514.890677  0101000020957F00005D19BF45294958C18FF7F8E260A8...   
132583   -396.937328  0101000020957F0000FBE7171F056F5AC1E8A6C0A5AF55...   
...              ...                                                ...   
135088   -595.319338  0101000020957F0000E14AD4DBC7765AC1F7B33C85B153...   
135104   -694.179039  0101000020957F00007CDF5EAFC58157C1645743B23E4F...   
135106  -1009.760928  0101000020957F0000649D6F21634858C119AE9BF528A3...   
135108  -1110.269437  0101000020957F00008FAE40D59E4B58C112C66046D597...   
135109   -396.941400  0101000020957F0000A6BF695F136F5AC1DADF87B20556...   

                                                      name  \
placeID                                                      
132560   puesto de gorditaspuesto de gorditaspuesto de ...   
132561            cafe ambarcafe ambarcafe ambarcafe ambar   
132564                        churchschurchschurchschurchs   
132572   Cafe ChairesCafe ChairesCafe ChairesCafe Chair...   
132583   McDonalds CentroMcDonalds CentroMcDonalds Cent...   
...                                                    ...   
135088   Cafeteria cenidetCafeteria cenidetCafeteria ce...   
135104                        vipsvipsvipsvipsvipsvipsvips   
135106   El Rincón de San FranciscoEl Rincón de San Fra...   

`Restaurant Name` Merging

Pertama, definisikan variabel all_resto_rate dengan variabel rating yang telah kita ketahui sebelumnya.

In [22]:
# Definisikan dataframe rating ke dalam variabel all_resto_rate
all_resto_rate = rating
all_resto_rate

userID  placeID  rating  food_rating  service_rating
0     U1077   135085       2            2               2
1     U1077   135038       2            2               1
2     U1077   132825       2            2               2
3     U1077   135060       1            2               2
4     U1068   135104       1            1               2
...     ...      ...     ...          ...             ...
1156  U1043   132630       1            1               1
1157  U1011   132715       1            1               0
1158  U1068   132733       1            1               0
1159  U1068   132594       1            1               1
1160  U1068   132660       0            0               0

[1161 rows x 5 columns]

Selanjutnya, untuk mengetahui nama restoran dengan placeID tertentu, mari kita gabungkan data geo yang berisikan placeID dan nama resto berdasarkan placeID dan assign ke variabel all_resto_name

In [23]:
# Menggabungkan all resto_rate dengan dataframe geo berdasarkan placeID
all_resto_name = pd.merge(all_resto_rate, geo[['placeID','name']], on='placeID', how='left')

# Print dataframe all_resto_name
all_resto_name

userID  placeID  rating  food_rating  service_rating  \
0     U1077   135085       2            2               2   
1     U1077   135038       2            2               1   
2     U1077   132825       2            2               2   
3     U1077   135060       1            2               2   
4     U1068   135104       1            1               2   
...     ...      ...     ...          ...             ...   
1156  U1043   132630       1            1               1   
1157  U1011   132715       1            1               0   
1158  U1068   132733       1            1               0   
1159  U1068   132594       1            1               1   
1160  U1068   132660       0            0               0   

                                       name  
0                    Tortas Locas Hipocampo  
1                     Restaurant la Chalita  
2                           puesto de tacos  
3                   Restaurante Marisco Sam  
4                                      vips  
...                                     ...  
1156                             palomo tec  
1157                   tacos de la estacion  
1158                          Little Cesarz  
1159     tacos de barbacoa enfrente del Tec  
1160  carnitas mata calle Emilio Portes Gil  

[1161 rows x 6 columns]

`Restaurant and Cuisine` Merging

Langkah selanjutnya adalah menggabungkan variabel all_resto_name yang kita peroleh dari tahapan sebelumnya dengan fitur masakan resto (cuisine).

In [24]:
# Menggabungkan dataframe cuisine dengan all_resto_name dan memasukkannya ke dalam variabel all_resto
all_resto = pd.merge(all_resto_name, cuisine, on='placeID', how='left')
all_resto

userID  placeID  rating  food_rating  service_rating  \
0     U1077   135085       2            2               2   
1     U1077   135038       2            2               1   
2     U1077   132825       2            2               2   
3     U1077   135060       1            2               2   
4     U1068   135104       1            1               2   
...     ...      ...     ...          ...             ...   
1326  U1043   132630       1            1               1   
1327  U1011   132715       1            1               0   
1328  U1068   132733       1            1               0   
1329  U1068   132594       1            1               1   
1330  U1068   132660       0            0               0   

                                       name   Rcuisine  
0                    Tortas Locas Hipocampo  Fast_Food  
1                     Restaurant la Chalita        NaN  
2                           puesto de tacos    Mexican  
3                   Restaurante Marisco Sam    Seafood  
4                                      vips    Mexican  
...                                     ...        ...  
1326                             palomo tec    Mexican  
1327                   tacos de la estacion    Mexican  
1328                          Little Cesarz   Pizzeria  
1329     tacos de barbacoa enfrente del Tec    Mexican  
1330  carnitas mata calle Emilio Portes Gil        NaN  

[1331 rows x 7 columns]

## Data Preparation

### Taking care on Missing Values

In [25]:
# Mengecek missing value pada dataframe all_resto
all_resto.isnull().sum()

userID              0
placeID             0
rating              0
food_rating         0
service_rating      0
name                0
Rcuisine          288
dtype: int64

 Untuk saat ini kita akan drop saja missing value ini.

In [26]:
# Membersihkan missing value dengan fungsi dropna()
all_resto_clean = all_resto.dropna()
all_resto_clean

userID  placeID  rating  food_rating  service_rating  \
0     U1077   135085       2            2               2   
2     U1077   132825       2            2               2   
3     U1077   135060       1            2               2   
4     U1068   135104       1            1               2   
5     U1068   132740       0            0               0   
...     ...      ...     ...          ...             ...   
1325  U1043   132732       1            1               1   
1326  U1043   132630       1            1               1   
1327  U1011   132715       1            1               0   
1328  U1068   132733       1            1               0   
1329  U1068   132594       1            1               1   

                                    name   Rcuisine  
0                 Tortas Locas Hipocampo  Fast_Food  
2                        puesto de tacos    Mexican  
3                Restaurante Marisco Sam    Seafood  
4                                   vips    Mexican  
5          Carreton de Flautas y Migadas    Mexican  
...                                  ...        ...  
1325                  Taqueria EL amigo     Mexican  
1326                          palomo tec    Mexican  
1327                tacos de la estacion    Mexican  
1328                       Little Cesarz   Pizzeria  
1329  tacos de barbacoa enfrente del Tec    Mexican  

[1043 rows x 7 columns]

Let's Recheck

In [27]:
# Mengecek kembali missing value pada variabel all_resto_clean
all_resto_clean.isnull().sum()

userID            0
placeID           0
rating            0
food_rating       0
service_rating    0
name              0
Rcuisine          0
dtype: int64

Merging and Formatting `Cuisine`

Pertama, mari kita cek ulang data setelah proses cleaning pada tahap sebelumnya. Kita buat variabel baru bernama fix_resto untuk menyimpan dataframe.

In [28]:
# Mengurutkan resto berdasarkan PlaceID kemudian memasukkannya ke dalam variabel fix_resto
fix_resto = all_resto_clean.sort_values('placeID', ascending=True)
fix_resto

userID  placeID  rating  food_rating  service_rating  \
1288  U1050   132560       0            2               0   
42    U1082   132560       0            0               0   
1303  U1087   132560       1            2               1   
14    U1067   132560       1            0               0   
854   U1092   132572       1            0               0   
...     ...      ...     ...          ...             ...   
189   U1055   135106       2            2               2   
1071  U1041   135109       1            2               1   
99    U1030   135109       0            0               0   
178   U1020   135109       2            2               1   
180   U1051   135109       1            1               1   

                            name   Rcuisine  
1288          puesto de gorditas   Regional  
42            puesto de gorditas   Regional  
1303          puesto de gorditas   Regional  
14            puesto de gorditas   Regional  
854                 Cafe Chaires  Cafeteria  
...                          ...        ...  
189   El Rincón de San Francisco    Mexican  
1071                   Paniroles    Italian  
99                     Paniroles    Italian  
178                    Paniroles    Italian  
180                    Paniroles    Italian  

[1043 rows x 7 columns]

In [29]:
# Mengecek berapa jumlah fix_resto
len(fix_resto.placeID.unique())

95

In [30]:
# Mengecek kategori masakan yang unik
fix_resto.Rcuisine.unique()

array(['Regional', 'Cafeteria', 'American', 'Mexican', 'Fast_Food',
       'Italian', 'Armenian', 'Pizzeria', 'Japanese', 'Vietnamese',
       'Family', 'International', 'Game', 'Burgers', 'Bakery', 'Bar',
       'Breakfast-Brunch', 'Bar_Pub_Brewery', 'Mediterranean',
       'Cafe-Coffee_Shop', 'Contemporary', 'Seafood', 'Chinese'],
      dtype=object)

Salah satu Cuisine yang menarik disini adalah `Game`, lets dive it more

In [31]:
# Mengecek kategori masakan Game
fix_resto[fix_resto['Rcuisine'] == 'Game']

userID  placeID  rating  food_rating  service_rating name Rcuisine
781   U1015   132851       1            1               1  KFC     Game
708   U1008   132851       1            1               1  KFC     Game
509   U1052   132851       1            0               2  KFC     Game
1188  U1131   132851       2            2               2  KFC     Game
770   U1037   132851       2            2               1  KFC     Game
764   U1111   132851       2            1               0  KFC     Game
574   U1069   132851       1            0               0  KFC     Game

Okay thats from KFC, mungkin makanan dengan hadiah? Let's Check any data containing `KFC` dan kita cek `Rcuisine`

In [32]:
# Mengecek kategori masakan pada nama restoran KFC
fix_resto[fix_resto['name'] == 'KFC']

userID  placeID  rating  food_rating  service_rating name  Rcuisine
573   U1069   132851       1            0               0  KFC  American
781   U1015   132851       1            1               1  KFC      Game
708   U1008   132851       1            1               1  KFC      Game
780   U1015   132851       1            1               1  KFC  American
763   U1111   132851       2            1               0  KFC  American
508   U1052   132851       1            0               2  KFC  American
509   U1052   132851       1            0               2  KFC      Game
1188  U1131   132851       2            2               2  KFC      Game
1187  U1131   132851       2            2               2  KFC  American
769   U1037   132851       2            2               1  KFC  American
770   U1037   132851       2            2               1  KFC      Game
707   U1008   132851       1            1               1  KFC  American
764   U1111   132851       2            1               0  KFC      Game
574   U1069   132851       1            0               0  KFC      Game

KFC memiliki dua kategori masakan yang berbeda, `Game` dan `American`. Tentu ini harus diperbaiki.

In [33]:
# Mengubah nama kategori masakan ‘Game’ menjadi ‘American’
fix_resto = fix_resto.replace('Game', 'American')
fix_resto[fix_resto['name'] == 'KFC']

userID  placeID  rating  food_rating  service_rating name  Rcuisine
573   U1069   132851       1            0               0  KFC  American
781   U1015   132851       1            1               1  KFC  American
708   U1008   132851       1            1               1  KFC  American
780   U1015   132851       1            1               1  KFC  American
763   U1111   132851       2            1               0  KFC  American
508   U1052   132851       1            0               2  KFC  American
509   U1052   132851       1            0               2  KFC  American
1188  U1131   132851       2            2               2  KFC  American
1187  U1131   132851       2            2               2  KFC  American
769   U1037   132851       2            2               1  KFC  American
770   U1037   132851       2            2               1  KFC  American
707   U1008   132851       1            1               1  KFC  American
764   U1111   132851       2            1               0  KFC  American
574   U1069   132851       1            0               0  KFC  American

Now, the data has been cleaned, mari kita Prepared datanya!

In [34]:
# Membuat variabel preparation yang berisi dataframe fix_resto kemudian mengurutkan berdasarkan placeID
preparation = fix_resto
preparation.sort_values('placeID')

userID  placeID  rating  food_rating  service_rating  \
1288  U1050   132560       0            2               0   
42    U1082   132560       0            0               0   
14    U1067   132560       1            0               0   
1303  U1087   132560       1            2               1   
1055  U1061   132572       1            2               2   
...     ...      ...     ...          ...             ...   
189   U1055   135106       2            2               2   
1071  U1041   135109       1            2               1   
99    U1030   135109       0            0               0   
178   U1020   135109       2            2               1   
180   U1051   135109       1            1               1   

                            name   Rcuisine  
1288          puesto de gorditas   Regional  
42            puesto de gorditas   Regional  
14            puesto de gorditas   Regional  
1303          puesto de gorditas   Regional  
1055                Cafe Chaires  Cafeteria  
...                          ...        ...  
189   El Rincón de San Francisco    Mexican  
1071                   Paniroles    Italian  
99                     Paniroles    Italian  
178                    Paniroles    Italian  
180                    Paniroles    Italian  

[1043 rows x 7 columns]

### Dropping Duplicates

In [35]:
# Membuang data duplikat pada variabel preparation
preparation = preparation.drop_duplicates('placeID')
preparation

userID  placeID  rating  food_rating  service_rating  \
1288  U1050   132560       0            2               0   
854   U1092   132572       1            0               0   
27    U1044   132583       1            2               1   
148   U1063   132584       2            1               1   
1304  U1087   132594       1            2               1   
...     ...      ...     ...          ...             ...   
1016  U1019   135086       0            0               0   
179   U1051   135088       1            1               1   
4     U1068   135104       1            1               2   
807   U1135   135106       0            0               0   
1071  U1041   135109       1            2               1   

                                    name   Rcuisine  
1288                  puesto de gorditas   Regional  
854                         Cafe Chaires  Cafeteria  
27                      McDonalds Centro   American  
148                   Gorditas Dona Tota    Mexican  
1304  tacos de barbacoa enfrente del Tec    Mexican  
...                                  ...        ...  
1016         Mcdonalds Parque Tangamanga    Burgers  
179                    Cafeteria cenidet  Cafeteria  
4                                   vips    Mexican  
807           El Rincón de San Francisco    Mexican  
1071                           Paniroles    Italian  

[95 rows x 7 columns]

### Data Conversion

In [36]:
# Mengonversi data series ‘placeID’ menjadi dalam bentuk list
resto_id = preparation['placeID'].tolist()

# Mengonversi data series ‘Name’ menjadi dalam bentuk list
resto_name = preparation['name'].tolist()

# Mengonversi data series ‘Rcuisine’ menjadi dalam bentuk list
resto_cuisine = preparation['Rcuisine'].tolist()

print(len(resto_id))
print(len(resto_name))
print(len(resto_cuisine))

95
95
95


### Dictionary Making

Kita akan membuat dictionary untuk menentukan pasangan key-value pada data resto_id, resto_name, dan resto_cuisine yang telah kita siapkan sebelumnya.

In [37]:
# Membuat dictionary untuk data ‘resto_id’, ‘resto_name’, dan ‘cuisine’
resto_new = pd.DataFrame({
    'id': resto_id,
    'resto_name': resto_name,
    'cuisine': resto_cuisine
})
resto_new

id                          resto_name    cuisine
0   132560                  puesto de gorditas   Regional
1   132572                        Cafe Chaires  Cafeteria
2   132583                    McDonalds Centro   American
3   132584                  Gorditas Dona Tota    Mexican
4   132594  tacos de barbacoa enfrente del Tec    Mexican
..     ...                                 ...        ...
90  135086         Mcdonalds Parque Tangamanga    Burgers
91  135088                   Cafeteria cenidet  Cafeteria
92  135104                                vips    Mexican
93  135106          El Rincón de San Francisco    Mexican
94  135109                           Paniroles    Italian

[95 rows x 3 columns]